Added cases in permutedims for sparse arrays

kangzhiq opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
#16941 `permutedims `and `transpose`

#### Brief description of what is fixed or changed
Fixed issue about casting sparse array to list in `permutedims` as well as `transpose`, since the later one call implicitly `permutedims(Array, (1, 0))` for evaluation.
Before:
```python
>>> b = MutableSparseNDimArray.zeros(20000, 10000, 20000, 4000)
>>> permutedims(b, (0, 2))
MemoryError
```
now:
```python
>>> A = MutableSparseNDimArray.zeros(20000, 10000, 20000, 4000)
>>> permutedims(A, (0, 1, 2)) == A
```

#### Other comments
[Don't merge] please.
Some tests would fail if we include sparse array into the test loop.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * Added specific cases for sparse array in `permutedims `and tests
<!-- END RELEASE NOTES -->

Commit:
Added cases in permutedims for spar_arr

- Added cases in permutedims for sparse arrays
- Added tests accordingly

kangzhiq commented:
The error on the test:
```CondaHTTPError: HTTP 504 GATEWAY_TIMEOUT for url <https://conda.anaconda.org/conda-forge/label/cf201901/linux-64/m4ri-20140914-h6debe1e_0.tar.bz2>
Elapsed: 00:59.228649
CF-RAY: 4e78e2cb7fb29d8f-ORD
An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.```

Commit:
Ameliorated tests

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -279,6 +281,16 @@ def permutedims(expr, perm):
+            index = 0
+            for i, idx in enumerate(perm(expr._get_tuple_index(k))):
+               index = index * new_shape[i] + idx
+            new_array[index] = v
```
are you sure you can't do this in one line? I believe you can use a dictionary comprehension.

kangzhiq replied:
I am afraid not. the problem is in the inner loop. The inner loop sums up terms in a succesive way, so I couldn't see how dict comprehension can be used in this case. Even `Sum()` is not applicable here I think.
In fact the inner loop have the same functionality as `_parse_index` of NDimArray. One alternative solution would be:
```python
# Initializing a mutable sparse array at first
temp_arr = MutableSparseNDimArray({}, new_shape)
temp_arr._sparse_array = {temp_arr._parse_index(perm(expr._get_tuple_index(k))):v 
                              for k,v in expr._sparse_array.items()}
if isinstance(expr, Immutable):
    return temp_arr.as_immutable()
return temp_arr
```
But the code is not very readable and it will create an unnecessary new `mutable` object if expr is `immutable`
What do you think of it?

Upabjojr replied:
what about:

```python
return type(expr)(
    {temp_arr._parse_index(perm(expr._get_tuple_index(k))): v for k, v in expr._sparse_array.items()},
    new_shape
)
```

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -279,6 +281,16 @@ def permutedims(expr, perm):
+    new_shape = perm(expr.shape)
+
+    if isinstance(expr, SparseNDimArray):
+        new_array = {}
```
call it `new_dict`, it's not an array.

kangzhiq replied:
I was calling it `new_array` to respect the existing convention. Some lines below there is a list being called as `new_array` and then passed to the constructor. I think the code might be more comprehensible if we keep that convention.

Upabjojr replied:
Which convention are you talking about? Suppose you were to write the code in C++, how would you call the variables?

kangzhiq replied:
I was trying to use the same name for the same variable in one function, I thought it could help improve the coherence within the function. Anyway there is no longer this problem.

Upabjojr replied:
No, it's confusing if you read the code. You are calling a dictionary an array, that's wrong.

kangzhiq replied:
Ok, I got your point. I will keep this in mind. :-)

Upabjojr commented on sympy/tensor/array/tests/test_arrayop.py:
```diff
@@ -277,3 +277,11 @@ def test_array_permutedims():
 
     assert permutedims(po2, (3, 2, 0, 1)) == Array([[[[sa[0], sa[4], sa[8], sa[12], sa[16], sa[20], sa[24], sa[28], sa[32]], [sa[36], sa[40], sa[44], sa[48], sa[52], sa[56], sa[60], sa[64], sa[68]], [sa[72], sa[76], sa[80], sa[84], sa[88], sa[92], sa[96], sa[100], sa[104]], [sa[108], sa[112], sa[116], sa[120], sa[124], sa[128], sa[132], sa[136], sa[140]]], [[sa[2], sa[6], sa[10], sa[14], sa[18],  ...
+
+    # test for large scale sparse array
```
also test the existing tests both on sparse and dense arrays.

kangzhiq replied:
Some tests will fail after including sparse array.  The error message is:
```
AttributeError: 'Tuple' object has no attribute 'as_coeff_Mul'
```
I will try to solve them.

kangzhiq replied:
Problem solved! As well as the bug of reshaping sparse array.

Commit:
Updated permutedims

- Simplified codes in permutedims to one line
- Added cases for dict of sparse array in _handle_ndarray_creation_inputs

Commit:
Added test and fixed bug of reshape sparse array

- Added test for both dense and sparse array in permutedims
- Fixed bug of reshaping a sparse array

Commit:
Updated tests
